### PR TITLE
WebCL: Move WebCLException to core/webcl.

### PR DESCRIPTION
**Pull request by rakuco**
This solves a layering violation, even if in an ugly way: `ExceptionState` depends on `V8ThrowException`, which was including and depending on `WebCLException`. The former is in `bindings/core`, which is not supposed to depend on anything in `modules/`, which is where the latter was.

The WebCLException.* files do not have any dependency on other files in `modules/webcl` (some unnecessary includes have been removed), so can easily move them to `core/webcl` instead and resolve the layering violations.

The deeper issue of whether `WebCLException` should inherit from `DOMException` and how to do this cleanly in Blink persists.

This change is a requirement for moving to M45, as work on upstream [bug 358074](https://code.google.com/p/chromium/issues/detail?id=358074) has reinforced these layering separations and causing code built with `component=shared_library` to fail to link.

### CODE REVIEW

**Comment by rakuco** [comment]
ping @ds-hwang @hujiajie @chunkyan

File: Source/core/webcl/WebCLException.cpp [hunk 1]
```diff
@@ -7,9 +7,6 @@
 
 #if ENABLE(WEBCL)
 
-#include "modules/webcl/WebCL.h"
-#include "modules/webcl/WebCLException.h"
```
Comment (hujiajie) [inline Source/core/webcl/WebCLException.cpp]: #include "core/webcl/WebCLException.h"?

File: Source/core/webcl/WebCLException.cpp [hunk 2]
```diff
@@ -7,9 +7,6 @@
 
 #if ENABLE(WEBCL)
 
-#include "modules/webcl/WebCL.h"
```
Comment (hujiajie) [inline Source/core/webcl/WebCLException.cpp]: I think this is for the definition of CL_* in this file, though WebCL.h is not the exact file that should be included here. The original header inclusion is a little messy, so I'd like to fix that later.

File: Source/modules/webcl/WebCLExtension.h [hunk 3]
```diff
@@ -9,7 +9,7 @@
 #if ENABLE(WEBCL)
 
 #include "modules/webcl/WebCLConfig.h"
-#include "modules/webcl/WebCLException.h"
+#include "core/webcl/WebCLException.h"
```
Comment (hujiajie) [inline Source/modules/webcl/WebCLExtension.h]: Will it break the build if it is put before WebCLConfig.h?

**Comment by hujiajie** [comment]
I cannot agree more that WebCLException should not exist in Blink at all.

https://codereview.chromium.org/17229006
https://codereview.chromium.org/17761003
https://codereview.chromium.org/17919002

The only reason for its existence is that the WebCL spec is influenced greatly by the legacy WebKit and Firefox idl style, and the conformance test (https://github.com/KhronosGroup/WebCL-conformance)  depends heavily on WebCLException instead of DOMException. @ds-hwang @chunkyan, do we really need to follow the spec strictly here, or deprecate WebCLException in the future and patch the conformance test to use additional WebCL DOMExceptions instead?

**Comment by rakuco** [comment]
thanks for the insightful comments; it turns out I... forgot to test an Android build with this change :-) I only verified that Linux and ozone-wayland were building again in M45 but didn't check if Android, which is the only one that actually enables WebCL, was also building fine.
I'll have to do some more changes and send another version of the patch.

**Comment by rakuco** [comment]
@hujiajie: I've tested an M45 Android build of patch v3 and it worked.

I've fixed some include problems: some headers in `modules/webcl` depend on `CL/cl.h` but were not directly including it, and I've worked around the possible dependency WebCLException would have on `WebCLConfig.h` by including `CL/cl.h` there to get all the macros it uses.

As for the WebCLException vs DOMException debate: do we have any contact with the people working on the spec? It's not clear to me if it is still being worked on, and I didn't find it very helpful. WebCLException's IDL in the spec says it inherits from DOMException, but none of the implementations (ours, Samsung's or [Nokia's](https://github.com/toaarnio/webcl-firefox)) seem to do that.

**Comment by chunkyan** [comment]
I agree with @hujiajie  about WebCLException should not exist in Blink,too. The direct impact of importing WebCLException class as a new type exception in blink is that: other classes which are not related to WebCL could also throw WebCLException , it 's really strange. 
and as @rakuco said, in the spec the definition of WebCLException is inherits from DOMException. 
So I think the implementation about WebCLException judgement is not good in webcl  conformance-test .
I think deprecate WebCLException in the future and patch the conformance test to use additional WebCL DOMExceptions instead is a good solution.

**Comment by hujiajie** [comment]
The latest patch LGTM, thanks @rakuco 
On WebCLException, I'll follow it up further and I'm willing to hear opinions from our side.

**Pull request merged**